File: 백준/Silver/2512. 예산/예산.py
import sys
N = int(input())
values = list(map(int,sys.stdin.readline().split()))
M = int(input())

l, r = 0,max(values)

while(l<=r):
  mid = (l+r)//2
  result = 0
  for value in values:
    if value <= mid:
      result+=value
    else:
      result+=mid
  if result>M:
    r = mid-1
  else:
    l = mid+1

print(r)

File: 백준/Silver/2512. 예산/예산.js
const fs = require('fs')
const input = fs.readFileSync('/dev/stdin','utf8').trim().split('\n')

const N = Number(input.shift())
const values = input.shift().trim().split(' ').map(Number)
const M = Number(input.shift())

let start = 0
let end = Math.max(...values)

while(start <= end){
    let mid = Math.floor((start + end)/2)
    let result = 0
    for (let value of values){
        if (value <= mid){result+=value}
        else {result+=mid}
    }
    if (result > M){
        end = mid-1
    } else {
        start = mid+1
    }
}

console.log(end)